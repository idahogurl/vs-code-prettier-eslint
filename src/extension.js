/* eslint-disable import/no-unresolved */
/**
 * @typedef {import('vscode')}
 */

import {
  languages, window, TextEdit, Range, workspace,
} from 'vscode';

import format from './formatter';
import { isFilePathMatchedByEslintIgnore, isFilePathMatchedByPrettierIgnore } from './ignore';

const path = require('path');

const rawResolve = require.resolve;

let outputChannel;

async function formatter(document) {
  const documentPath = path.dirname(document.fileName);
  const workspaceDir = workspace?.workspaceFolders.find((w) => documentPath.startsWith(w.uri.path))
    ?.uri.path;
  try {
    if (
      isFilePathMatchedByEslintIgnore(document.fileName, workspaceDir)
      || isFilePathMatchedByPrettierIgnore(document.fileName, workspaceDir)
      // eslint-disable-next-line no-empty
    ) {
      console.log('File ignored.');
      outputChannel.appendLine('File ignored. Matches entry in .eslintignore or .prettierignore');
    } else {
      const startLine = document.lineAt(0);
      const endLine = document.lineAt(document.lineCount - 1);
      const range = new Range(startLine.range.start, endLine.range.end);

      const text = document.getText(range);
      const extensionConfig = workspace?.getConfiguration('vs-code-prettier-eslint');

      /**
       * In some case, user cannot format his/her codes because of the error
       * "cannot find module @typescript-eslint/parser".
       *
       * In order to solve this problem, firstly take a try to search
       * "@typescript-eslint/parser" in node_modules of user's work directory.
       *
       * If find it successfully, intercept the `require.resolve` to redirect
       * the resolved path.
       */
      try {
        const typescriptEslintParserPath = require.resolve('@typescript-eslint/parser', {
          paths: [path.join(workspaceDir, 'temp.js')],
        });

        require.resolve = function resolve(...args) {
          const [moduleName] = args;
          if (moduleName && moduleName === '@typescript-eslint/parser') return typescriptEslintParserPath;
          return rawResolve(...args);
        };
      } catch (error) {
        /** do nothing, just follow default action of require.resolve * */
      }

      const formatted = await format({
        text,
        filePath: document.fileName,
        extensionConfig,
      });

      return [TextEdit.replace(range, formatted)];
    }
  } catch (err) {
    outputChannel.appendLine(`Error: ${err.message} \n${err.stack}`);
  } finally {
    require.resolve = rawResolve;
  }
}

const formattingProvider = {
  provideDocumentRangeFormattingEdits(document) {
    return formatter(document);
  },
};

const supportedLanguages = [
  'css',
  'graphql',
  'html',
  'javascript',
  'javascriptreact',
  'json',
  'jsonc',
  'less',
  'markdown',
  'mdx',
  'scss',
  'svelte',
  'typescript',
  'typescriptreact',
  'vue',
  'yaml',
];

supportedLanguages.forEach((language) => {
  languages.registerDocumentRangeFormattingEditProvider(language, formattingProvider);
});

// Create output channel for error logging
outputChannel = window.createOutputChannel('Prettier Eslint');
