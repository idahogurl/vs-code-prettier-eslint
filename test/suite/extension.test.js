jest.setTimeout(20000);

const path = require('path');
const fs = require('fs');
const helper = require('./helper');

require('../../dist/extension');

const basePath = path.join(__dirname, '..', 'fixtures');
const sourceFile = path.join(__dirname, '..', 'fixtures', 'test.js');

describe('Extension Test Suite', () => {
  vscode.window.showInformationMessage('Start all tests.');

  beforeAll(async () => {
    await helper.changeConfig('[javascript]', {
      'editor.defaultFormatter': 'rvest.vs-code-prettier-eslint',
    });
  });

  test('Formats document using .eslintrc', async () => {
    const content = fs.readFileSync(sourceFile).toString().replace('/* eslint-disable */\n', '');
    const filePath = `${basePath}/temp.js`;
    fs.writeFileSync(filePath, content, { overwrite: true });
    const document = await helper.openFile(filePath);
    await vscode.commands.executeCommand('editor.action.formatDocument');
    const formatted = document.getText();

    expect(formatted).toMatchSnapshot();
  });

  test('Format document using .prettierrc', async () => {
    const document = await helper.openFile(sourceFile);
    await vscode.commands.executeCommand('editor.action.formatDocument');
    const formatted = document.getText();

    expect(formatted).toMatchSnapshot();
  });
  afterAll(() => {
    helper.clearConfig('[javascript]');
  });
});
