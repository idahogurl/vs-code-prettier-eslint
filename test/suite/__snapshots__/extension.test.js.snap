// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Extension Test Suite Format document using .prettierrc 1`] = `
"/* eslint-disable */
const options = {
    first: "hello, world",
    second: true,
    internationalizing: true,
}

export { options }
"
`;

exports[`Extension Test Suite Formats document using .eslintrc 1`] = `
"const options = {
    first:              'hello, world',
    second:             true,
    internationalizing: true,
};

export { options };
"
`;

exports[`Extension Test Suite setting "prettierLast" makes prettier run last 1`] = `
"const options = {
    first:              'hello, world',
    second:             true,
    internationalizing: true,
};

export { options };
"
`;
